Fix `changed` signal emission in `Curve::set_point_offset`

aryan-11825114 opened the pull request:
Fixes #95752

**Before**

[before.webm](https://github.com/user-attachments/assets/c35895e8-f6bd-4350-a727-9ba3fc6dbe4b)

**After**

[after.webm](https://github.com/user-attachments/assets/4458fe35-4ee6-400a-8c00-75c02801e981)

aryan-11825114 committed:
fix changed signal emission in Curve::set_point_offset

aryan-11825114 commented:
Hey there, I was just wondering if this could be merged sooner, it would mean the world for a newbie contributor like me 😀

adamscott commented:
cc. @Geometror

passivestar commented:
Didn't find this PR before opening [my PR](https://github.com/godotengine/godot/pull/103840), but this PR breaks the curve if the point ends up at a different index after being dragged

aryan-11825114 commented:
> Didn't find this PR before opening [my PR](https://github.com/godotengine/godot/pull/103840), but this PR breaks the curve if the point ends up at a different index after being dragged and also when you try to undo such change (If you want to fix it I'll close my PR)

Yes, thank you, I have rebased and modified the P.R., it is now identical to yours

kleonc commented on editor/plugins/curve_editor_plugin.cpp:
```diff
@@ -549,11 +549,11 @@ void CurveEdit::set_point_position(int p_index, const Vector2 &p_pos) {
 	undo_redo->create_action(TTR("Modify Curve Point"));
-	undo_redo->add_do_method(*curve, "set_point_value", initial_grab_index, p_pos.y);
 	undo_redo->add_do_method(*curve, "set_point_offset", initial_grab_index, p_pos.x);
+	undo_redo->add_do_method(*curve, "set_point_value", p_index, p_pos.y);
 	undo_redo->add_do_method(this, "set_selected_index", p_index);
-	undo_redo->add_undo_method(*curve, "set_point_value", p_index, initial_grab_pos.y);
 	undo_redo->add_undo_method(*curve, "set_point_offset", p_index, initial_grab_pos.x);
+	undo_redo->add_undo_method(*curve, "set_point_value", initial_grab_index, initial_grab_pos.y);
```
This is not a correct fix, conceptually both old/new operation sequences are equivalent. Changing the offset might indeed change the index. Hence:
- If changing value before offset, then the same index can be used for both operations (old code).
- If changing offset before value, then a different/updated index needs to be used for value change (new code).

Both approaches are conceptually correct, which means the root cause of the bug is somewhere else.

---

And indeed the problem lies in `Curve::set_point_offset`:
https://github.com/godotengine/godot/blob/cae3d722a3abd6d93d84a84beccc65a78948653f/scene/resources/curve.cpp#L252-L266
It results in emitting `Curve.changed` signal:
- within `_remove_point(p_index)`,
- within `_add_point(Vector2(p_offset, p.position.y))`.

Emitting the signal twice is non-ideal and could be improved. The cause of the bug is that the last signal is emitted when not all data is updated: readded point's tangents are still zero, and tangents of neighbor points with auto-linear-tangents are not yet updated.

To fix the issue properly, the `Curve.changed` signal should be emitted at the end of this method, after all data is updated (and ideally shouldn't result in many emissions).

passivestar replied:
Could also add `Curve::set_point_offset_and_value()` to only emit once per mouse event in CurveEdit

aryan-11825114 replied:
I have added a default `dirty` argument to `_add_point()` and `_remove_point()` method  and set it to false in `_set_point_offset()`, this fixes the issue, let me know if this is the correct fix.

kleonc replied:
Yes, this looks like a fine solution.

kleonc replied:
> Could also add `Curve::set_point_offset_and_value()` to only emit once per mouse event in CurveEdit

Could be done as well, yes. But I'd that's for a potential separate PR, as it would be more of an enhancement (rather than a bugfix like this PR is).

kleonc commented on scene/resources/curve.h:
```diff
@@ -152,8 +152,9 @@ class Curve : public Resource {
 			real_t left_tangent = 0,
 			real_t right_tangent = 0,
 			TangentMode left_mode = TANGENT_FREE,
-			TangentMode right_mode = TANGENT_FREE);
-	void _remove_point(int p_index);
+			TangentMode right_mode = TANGENT_FREE,
+			bool dirty = true);
+	void _remove_point(int p_index, bool dirty = true);
```
- `dirty` name is a little confusing, sounds like we're passing whether it's already dirty. A more explicit `mark_dirty` seems more intuitive.
- `p_` prefix should be added to all of these parameters.
- Matching changes in the `.cpp`.

```cpp
			real_t p_left_tangent = 0,
			real_t p_right_tangent = 0,
			TangentMode p_left_mode = TANGENT_FREE,
			TangentMode p_right_mode = TANGENT_FREE,
			bool p_mark_dirty = true);
	void _remove_point(int p_index, bool p_mark_dirty = true);
```

---

Also while at it, you could add `p_` prefix for these (they're already like that in the `.cpp`):
https://github.com/godotengine/godot/blob/bb4ae4166f287b91c04cbc3c9122f5e498b6467a/scene/resources/curve.h#L78-L87

https://github.com/godotengine/godot/blob/bb4ae4166f287b91c04cbc3c9122f5e498b6467a/scene/resources/curve.h#L132

And change this to `p_index`:
https://github.com/godotengine/godot/blob/bb4ae4166f287b91c04cbc3c9122f5e498b6467a/scene/resources/curve.h#L129

aryan-11825114 replied:
Made the suggested changes.

kleonc reviewed (commented):
Overall looks good to me, left some nitpicks.

kleonc commented:
Regarding PR/commit title "Fix" would be better than "fixing":
> ~fixing~Fix `changed` signal emission in `Curve::set_point_offset`

Haven't thought you'd use my suggestion from https://github.com/godotengine/godot/pull/96296#discussion_r1986979848 as is. :upside_down_face:

kleonc reviewed (approved):
LGTM!

Repiteo commented:
Thanks!

akien-mga commented:
Cherry-picked for 4.4.1.
